File: ABC/299/D_Find_by_Query.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
int main(){
    int N; cin >> N;
    vector<int> S(N, 3);
    int l = 0, r = N-1, mid = (l + r) / 2;
    S[0] = 0;
    S[N-1] = 1;
    while (r - l > 1) {
        cout << "? " << mid+1 << endl;
        int num; cin >> num;
        S[mid] = num;
        if (abs(S[mid] - S[mid+1]) == 1) {
            cout << "! " << mid+1 << endl;
            return 0;
        }
        if (abs(S[mid-1] - S[mid]) == 1) {
            cout << "! " << mid << endl;
            return 0;
        }
        num == 1 ? r = mid : l = mid;
        mid = (l + r) / 2;
    }
    abs(S[mid] - S[mid+1]) == 1 ? cout << "! " << mid+1 << endl : cout << "! " << mid << endl;
    return 0;
}

File: ABC/250~299/299/D_Find_by_Query.py
N = int(input())
S = [3] * N
S[0] = 0
S[-1] = 1
left = 0
right = N-1
mid = (left + right) // 2
while right - left > 1:
    print('?', str(mid+1))
    num = int(input())
    S[mid] = num
    if abs(S[mid] - S[mid+1]) == 1:
        print('!', str(mid+1))
        exit()
    if abs(S[mid-1] - S[mid]) == 1:
        print('!', str(mid))
        exit()
    if num == 1:
        right = mid
    else:
        left = mid
    mid = (left + right) // 2

if abs(S[mid] - S[mid+1]) == 1:
    print('!', str(mid+1))
    exit()
if abs(S[mid-1] - S[mid]) == 1:
    print('!', str(mid))
    exit()